metals/src/main/scala/scala/meta/internal/metals/BuildServerConnection.scala

```diff
@@ -47,9 +58,16 @@ class BuildServerConnection private (
   // the name is set before when establishing conenction
   def name: String = initialConnection.socketConnection.serverName
 
+  def onReconnection(
+      index: BuildServerConnection => Future[Unit]
+  ): Unit = {
+    onReconnection.set(index)
+  }
+
   /** Run build/shutdown procedure */
   def shutdown(): Future[Unit] = connection.map { conn =>
     try {
+      isShuttingDown.set(true)
```

Comment:
by using `compareAndSet(false, true` you could guarantee here that the method will be called exactly once, hence making it idempotent.

Reply:
Good idea! Fixed it!